### ISSUE TITLE
Proposal: allow comparison operators in switch case labels C#

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#812]
@JFMous commented on [Sat May 14 2016](https://github.com/dotnet/roslyn/issues/11312)

It would be nice if the case labels in switch statement supported comparison operators, much like the Select Case statement in Visual Basic does.

Now, there are only two options for the switch labels:

> switch-label:
> case _constant-expression_ :
> default :

This could be extended to:

> case [< | > | <= | >=] _constant-expression_ [ to ] _constant-expression_ :
> default :

So you could write code as in this example:

```
int iq = DoIqTest();

switch (iq)
{
    case <= 69:
        ProcedureExtremelyLow();
        break;
    case 70 to 79:
        BorderlineProcedure();
        break;
    case 80 to 89:
        LowAverageProcedure();
        break;
    case 90 to 99:
    case 101 to 109:
        AverageProcedure();
        break;
    case 100:
        ExactlyMedianProcedure();
        break;
    case 110 to 119:
        HighAverageProcedure();
        break;
    case 120 to 129:
         SuperiorProcedure();
         break;
    case >= 130:
        VerySuperiorProcedure();
        break;
}

```

The 'to' keyword would be used to specify a range. In the statement `switch (value)`, `case x to y:` would be equivalent to the boolean expression `value >= x && value <= y`

---

@HaloFour commented on [Sat May 14 2016](https://github.com/dotnet/roslyn/issues/11312#issuecomment-219259783)

Case guards in pattern matching #206 will open this up to allowing for arbitrary conditions:

``` cs
int iq = DoIqTest();

switch (iq)
{
    case * when iq <= 69:
        ProcedureExtremelyLow();
        break;
    case * when iq <= 79:
        BorderlineProcedure();
        break;
    case * when iq <= 89
        LowAverageProcedure();
        break;
    case * when iq == 100:
        ExactlyMedianProcedure();
        break;
    case * when iq <= 109:
        AverageProcedure();
        break;
    case * when iq <= 119:
        HighAverageProcedure();
        break;
    case * when iq <= 129:
         SuperiorProcedure();
         break;
    default:
        VerySuperiorProcedure();
        break;
}
```

The wildcard pattern would match on anything but then the case guard would contain the specific conditions.  They would be evaluated in lexical order.

---

@alrz commented on [Sun May 15 2016](https://github.com/dotnet/roslyn/issues/11312#issuecomment-219293994)

would it make sense to make `case` optional or perhaps switch expression?

``` cs
switch {
  when id <= 69: ... break;
  ...
}
```

might be preferrable over if else.

---

@bondsbw commented on [Mon May 16 2016](https://github.com/dotnet/roslyn/issues/11312#issuecomment-219612182)

@alrz I like that somewhat, feels like a reasonable simplification.

But I feel this may be even more useful in pattern matching.  In match expressions, the roughly equivalent form might be:

``` c#
var x = match
{
    when iq <= 69: ProcedureExtremelyLow(),
    when iq <= 79: BorderlineProcedure(),
    when iq <= 89: LowAverageProcedure(),
    when iq <= 99 || (iq <= 109 && iq >= 101): AverageProcedure(),
    when iq == 100: ExactlyMedianProcedure(),
    when iq <= 119: HighAverageProcedure(),
    when iq <= 129: SuperiorProcedure(),
    default: VerySuperiorProcedure()
};
```

I like the possibilities but if `iq` is always the comparison, it really feels like it should be the subject of the match expression.  Something like the following feels more like pattern matching (although it would require a bit of rethinking what a relational expression is):

``` c#
var x = iq match
{
    when <= 69: ProcedureExtremelyLow(),
    when <= 79: BorderlineProcedure(),
    when <= 89: LowAverageProcedure(),
    when == 100: ExactlyMedianProcedure(),
    when <= 109: AverageProcedure(),
    when <= 119: HighAverageProcedure(),
    when <= 129: SuperiorProcedure(),
    default: VerySuperiorProcedure()
};
```

And I'm not sure I like changing

``` c#
when iq <= 99 || (iq <= 109 && iq >= 101):
```

into something like 

``` c#
when <= 99 || (<= 109 && >= 101):
```

---

@Unknown6656 commented on [Mon May 23 2016](https://github.com/dotnet/roslyn/issues/11312#issuecomment-220968107)

@bondsbw, @JFMous : I would like to see something like the following implemented with the `match`-pattern:

```
Func<int, string> func1, func2, func3, ...;

int iq = ...;
string result = iq match
{
    when <= 100: func1,
    when == 100: func2,
    when >= 100: func3,
    // ....
    default: funcx()
}(iq);
```

Meaning, the possibility to use the match-statement inside of expressions.

---

@HaloFour commented on [Mon May 23 2016](https://github.com/dotnet/roslyn/issues/11312#issuecomment-220969612)

@Unknown6656 The proposal is that `match` is an expression, so of course you'd be able to use it within other expressions.  You can think of it like a ternary op on steroids.

As for your specific example, there have been no proposed range patterns.  Nor has there been a proposal to allow for only case guards with an implied wildcard pattern.  So as of now your example would be:

``` cs
string result = iq match (
    case * when iq < 100: func1,
    case * when iq == 100: func2,
    case * when iq > 100: func3,
    case *: funcx
)(iq);
```

**Comment by DavidArno** [comment]
Rather than implementing this limited feature, could the team just finish implementing pattern matching, including `match` expressions and get that released, please?

**Comment by bondsbw** [comment]
@DavidArno Haha there is a thing called "priorities".

**Comment by DavidArno** [comment]
@bondsbw,

Exactly. They should prioritise finishing the half-complete pattern matching feature over adding new features like nullable reference types and those interface thingies 😉

**Comment by eyalsk** [comment]
@DavidArno I agree that `match` expression would be awesome to have sooner than later but don't know if this is the place to discuss priorities, it's a reasonable suggestion, at least in my opinion.

p.s. I wouldn't downvote something just due to priorities, not saying you did but just a side note. 😉

**Comment by bondsbw** [comment]
I feel like these partial infix expressions could be a more general feature.

```c#
iqs.Where(x => x < 100)
```
could be simplified to
```c#
iqs.Where(< 100)
```

**Comment by eyalsk** [comment]
@bondsbw Well, if you think that this feature should be generalized then I think that it warrant a new post.

**Comment by bondsbw** [comment]
I'll start a new post if there is interest.

**Comment by qaqz111** [comment]
The pattern match already can handle this for now:

            var a = 16;
            switch (a)
            {
                case int _ when a < 100:
                    Console.WriteLine(a);
                    break;
                case int _ when a < 200:
                    Console.WriteLine(a + a);
                    break;
            }

With `when` you can do a more complex play:

            var a = 16;
            switch (a)
            {
                case int _ when a < 100:
                    Console.WriteLine(a);
                    break;
                case int _ when IsOK(a):
                    Console.WriteLine(a + a);
                    break;
            }

        private bool IsOK(int a) => true; //<- replace with your own logic

**Comment by renatoprogramer** [comment]
**Very good proposal**

**Comment by gafter** [comment]
I'm championing this as the idea of having comparison-based pattern-matching forms.

**Comment by gafter** [comment]
@HaloFour Yes, that syntax is fine. Also `case > 10`. Not sure how it should work when the input is of type object, for example.

**Comment by bondsbw** [comment]
I'm curious if that syntax can be generalized into partial application.

**Comment by gafter** [comment]
@bondsbw Can you give an example? Remember these are patterns, not expressions.

For ranges, I expect the syntax would be something like `in 1 to 10`, for example

`case in 1 to 10:`
`if (x is in 1 to 10)`

etc.

**Comment by bondsbw** [comment]
@gafter This would be outside of pattern matching, outside of the scope of the current proposal but related.

The idea would be to open up expressions like I mentioned above:

```c#
iqs.Where(< 100)
```

Say we have a binary operator `x op y` with the function signature `Tx * Ty -> Tresult`.  Then `x op` would be a partial application with function signature `Ty -> Tresult`, and `op y` would be a partial application with function signature `Tx -> Tresult`.

To tie it all together, partial applications could be substituted for method groups with a compatible signature.  `< 100` has the signature `int -> bool` and is compatible with the `predicate` parameter of `Where`.

**Comment by bondsbw** [comment]
A different idea... allow patterns to be substituted for any method group with signature `T -> bool`.

Then we could have
```c#
iqs.Where(is < 100)
```
but also the much broader range of patterns:
```c#
iqs.Where(is in 120 to 140)

people.Where(is Student { Gpa: > 3.0 })
```

**Comment by bondsbw** [comment]
@svick `+` and `-` are the only ones that could be ambiguous (_I think_... please correct me).  In those cases it would have to prefer to interpret it as the unary operator, for the sake of BC.

I think I prefer the pattern approach, though it is limited to situations where the return value is `bool`.

**Comment by theunrepentantgeek** [comment]
Any operator with both a binary and unary form could trigger this ambiguity - which includes `*`, `&` and `|` as well.

Building on the pattern matching syntax we already have, the **when** syntax opens up every possible predicate in an extremely readable manner.

**Comment by Mattias-Viklund** [comment]
Id like to propose an extension to this.

Where you can do something like this:

`
int a = 3;
int b = 2;
int c = 1;

switch (a) {
    case (&c == c):
        Console.WriteLine("(a & c) == c");
        break;
    case (>> 1 == c)
        Console.WriteLine("(a >> 1) == c");
        break;

    case (-b == c)
        Console.WriteLine("(a - b) == c");
        break;

}
`

Might be silly though.

**Comment by Unknown6656** [comment]
@Mattias-Viklund your last switch case is a bit flawed, as it would be confusing whether you meant `a-b == c` or `-b == c` (both are currently valid expressions).

I would rather see a Haskell-like currying and partial application of operators, however, this would require currying in general to be introduced to C#. This would allow expressions such as:
```csharp
Func<int, int> = (>>2);
```
...which could be applied to your example above in a more general manner.

**Comment by zspitz** [comment]
Would this also work with strings, as in VB.NET?
```csharp
var s = GetArbitraryString();
switch (s)
    case >= "n":
        Console.WriteLine("Past the midpoint");
        break;
    case "m" to "n":
        Console.WriteLine("Starts with 'm');
        break;
}
```

**Comment by Unknown6656** [comment]
@zspitz : Strings would have to define comparison operators like they do in VB.NET.
So, as long as they are not defined, your code wouldn't work.

However, you could currently "solve" your code sample by replacing strings with chars:
```csharp
string s = ........;

switch (s[0])
    case >= 'n':
        Console.WriteLine("Past the midpoint");
        break;
    case 'm' to 'n':
        Console.WriteLine("Starts with 'm'");
        break;
}
```

**Comment by Thaina** [comment]
I would go against the `to` syntax. It still very ambiguous

And I would favor `and` and `or` syntax in pattern. So `1 to 10` would just become `>= 1 and <= 10`
